sync/src/types.rs

```diff
@@ -933,16 +921,67 @@ impl<CS: ChainStore> SyncSharedState<CS> {
 
     pub fn mark_as_known_tx(&self, hash: H256) {
         self.inflight_transactions().remove(&hash);
-        self.tx_filter().insert(hash, ());
+        self.tx_filter.lock().insert(hash, ());
     }
 
     pub fn already_known_tx(&self, hash: &H256) -> bool {
-        self.tx_filter().contains_key(hash)
+        self.tx_filter.lock().contains_key(hash)
     }
 
     // Return true when the block is that we have requested and received first time.
     pub fn new_block_received(&self, block: &Block) -> bool {
         self.write_inflight_blocks()
             .remove_by_block(block.header().hash())
     }
+
+    pub fn insert_inflight_proposals(&self, ids: Vec<ProposalShortId>) -> Vec<bool> {
+        let mut locked = self.inflight_proposals.lock();
+        ids.into_iter().map(|id| locked.insert(id)).collect()
+    }
+
+    pub fn remove_inflight_proposals(&self, ids: &[ProposalShortId]) -> Vec<bool> {
+        let mut locked = self.inflight_proposals.lock();
+        ids.iter().map(|id| locked.remove(id)).collect()
+    }
+
+    pub fn get_block_status(&self, block_hash: &H256) -> BlockStatus {
+        let mut locked = self.block_status_map.lock();
+        match locked.get(block_hash).cloned() {
+            Some(status) => status,
+            None => {
+                if self.shared.store().get_block_header(block_hash).is_some() {
+                    locked.insert(block_hash.clone(), BlockStatus::BLOCK_HAVE_MASK);
+                    BlockStatus::BLOCK_HAVE_MASK
+                } else {
+                    BlockStatus::UNKNOWN
+                }
+            }
+        }
+    }
+
+    pub fn insert_block_status(&self, block_hash: H256, status: BlockStatus) {
+        self.block_status_map.lock().insert(block_hash, status);
+    }
+
+    pub fn clear_get_block_proposals(&self) -> FnvHashMap<ProposalShortId, FnvHashSet<PeerIndex>> {
+        let mut locked = self.pending_get_block_proposals.lock();
+        let old = locked.deref_mut();
+        let mut ret = FnvHashMap::default();
+        swap(old, &mut ret);
+        ret
+    }
+
+    pub fn insert_get_block_proposals(&self, ids: Vec<ProposalShortId>, pi: PeerIndex) {
```

Comment:
I wonder why `pi: PeerIndex` is not the first argument.

Reply:
Placing `pi` in the first one is better! Let me change it right now.